src/JsonSchema/RefResolver.php

```diff
@@ -9,269 +9,177 @@
+            $newSchema = $this->resolveCached((string) $jsonPointer, $paths);
+            $this->unionSchemas($newSchema, $refSchema, $fileName, $paths);
         }
 
-        $this->resolve($schema->$propertyName, $sourceUri);
+        unset($schema->{'$ref'});
+        if (!$this->hasSubSchemas($schema)) {
+            foreach (get_object_vars($refSchema) as $prop => $value) {
+                $schema->$prop = $value;
+            }
+        } else {
+            $newSchema = new \stdClass();
+            foreach (get_object_vars($schema) as $prop => $value) {
+                $newSchema->$prop = $value;
+                unset($schema->$prop);
+            }
+            $schema->allOf = array($newSchema, $refSchema);
+        }
     }
 
     /**
-     * Look for the $ref property in the object.  If found, remove the
-     * reference and augment this object with the contents of another
-     * schema.
-     *
-     * @param object $schema    JSON Schema to flesh out
-     * @param string $sourceUri URI where this schema was located
+     * @param object $schema
+     * @return bool
      */
-    public function resolveRef($schema, $sourceUri)
+    private function hasSubSchemas($schema)
     {
-        $ref = '$ref';
-
-        if (empty($schema->$ref)) {
-            return;
-        }
-
-        $splitRef = explode('#', $schema->$ref, 2);
-
-        $refDoc = $splitRef[0];
-        $refPath = null;
-        if (count($splitRef) === 2) {
-            $refPath = explode('/', $splitRef[1]);
-            array_shift($refPath);
-        }
-
-        if (empty($refDoc) && empty($refPath)) {
-            // TODO: Not yet implemented - root pointer ref, causes recursion issues
-            return;
-        }
-
-        if (!empty($refDoc)) {
-            $refSchema = $this->fetchRef($refDoc, $sourceUri);
-        } else {
-            $refSchema = $this->rootSchema;
-        }
-
-        if (null !== $refPath) {
-            $refSchema = $this->resolveRefSegment($refSchema, $refPath);
+        foreach (array_keys(get_object_vars($schema)) as $propertyName) {
+            if (in_array($propertyName, $this->getReservedKeysWhichAreInFactSubSchemas())) {
+                return true;
+            }
         }
 
-        unset($schema->$ref);
-
-        // Augment the current $schema object with properties fetched
-        foreach (get_object_vars($refSchema) as $prop => $value) {
-            $schema->$prop = $value;
-        }
+        return false;
     }
 
     /**
-     * Set URI Retriever for use with the Ref Resolver
-     *
-     * @param UriRetriever $retriever
-     * @return $this for chaining
+     * @return string[]
      */
-    public function setUriRetriever(UriRetriever $retriever)
-    {
-        $this->uriRetriever = $retriever;
-
-        return $this;
-    }
-
-    protected function resolveRefSegment($data, $pathParts)
+    private function getReservedKeysWhichAreInFactSubSchemas()
```

Comment:
Interesting name here. Very matter-of-fact. I don't have a problem with it, just commenting on the choice of words/phrasing.